File: src/main/scala/kafkastreams/scalaexamples/KafkaStreamsApp.scala
package kafkastreams.scalaexamples

import java.util.Properties
import java.util.concurrent.TimeUnit

import org.apache.kafka.streams.{KafkaStreams, StreamsConfig, Topology}
import org.apache.kafka.streams.scala.StreamsBuilder

trait KafkaStreamsApp {

  def createTopology(builder: StreamsBuilder): Topology

  def start(applicationId: String): KafkaStreams = {
    val config = new Properties()
    config.put(StreamsConfig.APPLICATION_ID_CONFIG, applicationId)
    config.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:29092")

    val builder = new StreamsBuilder
    val topology = createTopology(builder)

    val streams = new KafkaStreams(topology, config)
    streams.start()

    Runtime.getRuntime.addShutdownHook(new Thread(() =>
      streams.close(10, TimeUnit.SECONDS)
    ))

    streams
  }
}

File: src/main/java/kafkastreams/javaexamples/KafkaStreamsApp.java
package kafkastreams.javaexamples;

import org.apache.kafka.streams.*;
import org.apache.kafka.streams.state.QueryableStoreTypes;
import org.apache.kafka.streams.state.ReadOnlyWindowStore;
import org.apache.kafka.streams.state.WindowStoreIterator;

import java.util.Properties;
import java.util.concurrent.TimeUnit;

abstract class KafkaStreamsApp {

    public abstract Topology createTopology(StreamsBuilder builder);

    public KafkaStreams start(String applicationId) {
        Properties config = new Properties();
        config.put(StreamsConfig.APPLICATION_ID_CONFIG, applicationId);
        config.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:29092");

        StreamsBuilder builder = new StreamsBuilder();
        Topology topology = createTopology(builder);

        KafkaStreams streams = new KafkaStreams(topology, config);
        streams.start();

        Runtime.getRuntime().addShutdownHook(new Thread(() ->
                streams.close(10, TimeUnit.SECONDS)
        ));

        return streams;
    }
}
